pyClarion/numdicts/ops.py

```diff
@@ -56,57 +57,337 @@ def set_by(
+
+    return (g * isclose(d, reduce_min(d)),)
+
+
+@register_op
+def merge(*ds: NumDict) -> NumDict:
+    if len(ds) == 0:
+        raise ValueError("Nothing to merge.")
+
+    if len(set.union(*map(set, ds))) < sum(map(len, ds)):
+        raise ValueError("NumDicts are not disjoint")
+
+    data = {}
+    for d in ds:
+        data.update(d)
+
+    if len(set((d.default for d in ds))) == 1:
+        default = ds[0].default
+    else:
+        default = None
+    value = NumDict(data, default=default)
+    record_call(merge, value, ds, {})
+    return value
+
+
+@register_grad(merge)
+def _grad_merge(grads: NumDict, *ds: NumDict) -> Tuple[NumDict, ...]:
+    return tuple(NumDict({k: grads[k] for k in d}, grads.default) for d in ds)
+
+# We consider a function diffable if
+#   1) it is an op
+#   2) it is a wrapper for a sequence of ops
+
+
+def by(
+    d: NumDict,
+    *,
+    reducer: Callable[[NumDict, Hashable], NumDict],
+    keyfunc: Callable[[Hashable], Hashable]
+) -> NumDict:
+    """
+    Reduce values in d grouped by keyfunc.
+
+    If reducer is diffable, this function is diffable.
+
+    :param d: The target numdict.
+    :param reducer: An op that maps all values of a numdict to a single key.
+    :param keyfunc: The grouping function; maps keys to keys.
+    """
+
+    keys = tuple(set(keyfunc(k) for k in d))
+    if keys:
+        selectors = ([x for x in d if keyfunc(x) == k] for k in keys)
+        groups = (keep(d, keys=s) for s in selectors)
+        return merge(*[reducer(g, key=k) for k, g in zip(keys, groups)])
+    else:  # if keys is empty merge will fail
+        return d
+
+
+# This is an op b/c only calls diffable ops
+def sum_by(d: NumDict, *, keyfunc: Callable[[Hashable], Hashable]) -> NumDict:
+    return by(d, reducer=reduce_sum, keyfunc=keyfunc)
+
+# This is an op b/c only calls diffable ops
+
+
+def max_by(d: NumDict, *, keyfunc: Callable[[Hashable], Hashable]) -> NumDict:
+    return by(d, reducer=reduce_max, keyfunc=keyfunc)
+
+# This is an op b/c only calls diffable ops
+
+
+def min_by(d: NumDict, *, keyfunc: Callable[[Hashable], Hashable]) -> NumDict:
+    return by(d, reducer=reduce_min, keyfunc=keyfunc)
 
 
 @register_op
-def max_by(d: D, *, keyfunc: Callable[..., Hashable], **kwds: Any) -> NumDict:
+def boltzmann(d: D, t: Union[float, int]) -> NumDict:
+    """(low < d[k] < high)
+    Construct a boltzmann distribution from d with temperature t.
+
+    If d has a default, the returned value will have a default of 0, and, if d
+    is empty, the return value will also be empty.
+    """
+    default = 0 if d.default is not None else None
+
+    if len(d) > 0:
+        x = d / t
+        x = x - reduce_max(x).default  # softmax(x) = softmax(x + c)
+        numerators = x.exp()
+        denominator = reduce_sum(numerators).default
+        value = with_default(numerators / denominator, default=default)
+        kwds = {"t": t}
+        record_call(boltzmann, value, (d,), kwds)
```

Comment:
On second look, something seems off about autodiff support here. 

All the functions used to calculate boltzmann when `len(d) > 0` are themselves differentiable ops (with the exception of `with_default`). So, those ops will be on the gradient tape in addition to boltzmann itself, which is quite bad. Directly accessing the defaults on 267 and 269 blocks gradients from these ops from flowing to upstream NumDicts, I think. But, why not let the system do what it was made to do?

The main issue seems to be that pesky if-else. My original decision to support the case when `len(d) == 0` seems quite ill-advised in hindsight. My inclination now is to remove support for this case and let GradientTape take care of the rest. This will break various pyClarion components, but not so many as to be onerous. We will also have to sort out what to do about `with_default`. The best thing to do is probably to make it an op, that shouldn't be too hard.

I'm afraid that's the best we can do quickly, short of sending `boltzmann` back to funcs for now.